lang/java/reef-io/src/main/java/org/apache/reef/io/network/shuffle/impl/StaticShuffleClient.java

```diff
@@ -0,0 +1,204 @@
+  private final LinkListener<Message<ShuffleControlMessage>> controlLinkListener;
+
+  @Inject
+  public StaticShuffleClient(
+      final ShuffleDescriptor initialShuffleDescriptor,
+      final ClientTupleCodecMap tupleCodecMap,
+      final TupleOperatorFactory tupleOperatorFactory) {
+
+    this.initialShuffleDescriptor = initialShuffleDescriptor;
+    this.tupleOperatorFactory = tupleOperatorFactory;
+    this.tupleCodecMap = tupleCodecMap;
+    this.shuffleMessageDispatcher = new ShuffleMessageDispatcher();
+    this.controlMessageHandler = new ControlMessageHandler();
+    this.controlLinkListener = new ControlLinkListener();
+    this.groupingSetupLatchMap = new ConcurrentHashMap<>();
+    for (final String groupingName : initialShuffleDescriptor.getGroupingNameList()) {
+      groupingSetupLatchMap.put(groupingName, new CountDownLatch(1));
+    }
+
+    this.clientSetupLatch = new CountDownLatch(groupingSetupLatchMap.size());
+  }
+
+  @Override
+  public ShuffleDescriptor getShuffleDescriptor() {
+    return initialShuffleDescriptor;
+  }
+
+  @Override
+  public void waitForGroupingSetup(final String groupingName) {
+    try {
+      groupingSetupLatchMap.get(groupingName).await();
+    } catch (final InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void waitForSetup() {
+    try {
+      clientSetupLatch.await();
+    } catch (final InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private synchronized void groupingInitialized(final String groupingName) {
+    if (groupingSetupLatchMap.get(groupingName).getCount() != 0) {
+      groupingSetupLatchMap.get(groupingName).countDown();
+      clientSetupLatch.countDown();
+    }
+  }
+
+  private void onDriverMessage(final ShuffleControlMessage driverMessage) {
+    if (driverMessage.getCode() == StaticShuffleMessageCode.GROUPING_SETUP) {
+      groupingInitialized(driverMessage.getGroupingName());
+    }
+  }
+
+  @Override
+  public EventHandler<Message<ShuffleTupleMessage>> getTupleMessageHandler() {
+    return shuffleMessageDispatcher.getTupleMessageHandler();
+  }
+
+  @Override
+  public LinkListener<Message<ShuffleTupleMessage>> getTupleLinkListener() {
+    return shuffleMessageDispatcher.getTupleLinkListener();
+  }
+
+  @Override
+  public Codec<Tuple> getTupleCodec(String groupingName) {
+    return tupleCodecMap.getTupleCodec(groupingName);
+  }
+
+  @Override
+  public <K, V> TupleReceiver<K, V> getReceiver(final String groupingName) {
+    return tupleOperatorFactory.newTupleReceiver(initialShuffleDescriptor.getGroupingDescriptor(groupingName));
+  }
+
+  @Override
+  public <K, V> TupleSender<K, V> getSender(final String groupingName) {
+    return tupleOperatorFactory.newTupleSender(initialShuffleDescriptor.getGroupingDescriptor(groupingName));
+  }
+
+  @Override
+  public <K, V> void registerTupleLinkListener(
```

Comment:
why all of the registering methods are separate? why don't you register all of the elements (listener, handler, ...) in a single method?